cirq-core/cirq/sim/operation_target.py

```diff
@@ -68,7 +68,7 @@ def apply_operation(self, op: 'cirq.Operation'):
         protocols.act_on(op, self)
 
     @abc.abstractmethod
-    def copy(self: TSelfTarget) -> TSelfTarget:
+    def copy(self: TSelfTarget, reuse_buffer: bool = False) -> TSelfTarget:
```

Comment:
Make sure to add Args to the docstring with an explanation of `reuse_buffer`. (Not just "True to reuse the buffer". Something like "If True, any buffers will be reused by the copy. This will save time by avoiding a deep copy of the buffers, but should only be used when there is no chance that the two objects will be writing to the buffers simultaneously.")

Reply:
I have added the paragraph to the docstring, but I do not fully understand it. It seems that there are no buffers shared between the iterations of the simulation.

https://github.com/quantumlib/Cirq/blob/b60b9f8fa285def55af608f47955f3c8c3e41687/cirq-core/cirq/sim/simulator_base.py#L268-L278

I cannot find where the shared buffer is stored. It is not in `ActOnDensityMatrixArgs` because not copying the buffer does not affect the simulation. And it is not in `DensityMatrixSimulator` either.

Reply:
Is this comment old? You have updated the function in `simulator_base` to apply the `resuse_buffers=True` option.

Reply:
We have not discussed it yet and I still hold the same opinion. I agree that without copying buffers would improve performance. I am just not sure about the name `reuse_buffer` and the explanation in the docstring.

Reply:
Sure, how would you phrase it? My wording probably wasn't the best.

Reply:
Can we say `skip_buffer`? The docstring can be something like `True to skip copying buffers`. I think we can just describe the *behavior* related to the parameter instead of the *purpose*.

Reply:
Hmm, that sounds to me like it would just set the new object buffer to null. But I think I understand your point in that reuse_buffers could be confusing too. Maybe something more direct like shallow_copy_buffers?

Reply:
Is the buffer shallow copied? I write only
```
        if not reuse_buffer:
            target.available_buffer = [b.copy() for b in self.available_buffer]
```
in `act_on_density_matrix_args.py`

Reply:
Well by "shallow copied", I mean the reference to `available_buffer` is copied, which occurs in `ActOnArgs.copy`. So documentation-wise, I guess it depends on whether you're viewing it from the perspective of the user who is calling `copy` or the perspective of the user who is implementing a new subclass of `ActOnArgs`.

Maybe clearer in both situations would be to call it `deep_copy_buffers`, and default that to True. I always prefer naming booleans such that the default is False, but perhaps it's more clear to use `deep_copy_buffers=True` as default here. (There are plenty of other places where we default a boolean to True, so it wouldn't be an oddball or anything). WDYT?

Reply:
Agree. I will update the code soon.